rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,71 @@
+---
+authors: Andrey Bulgakov (andrey.bulgakov@goteleport.com), Edoardo Spadolini (edoardo.spadolini@goteleport.com)
+state: draft
+---
+
+# RFD 0138 - Postgres Backend
+
+## What
+
+A Teleport backend is an abstraction over a storage system (ranging from SQLite on a local file to managed services like Amazon DynamoDB and Google Firestore), providing a key-value to be used by the Teleport auth servers to store all persistent cluster state. This RFD proposes a backend implementation to use [PostgreSQL](https://www.postgresql.org/) as the underlying storage mechanism.
+
+## Why
+
+Currently, users looking to self-host Teleport in a High Availability (HA) deployment outside of AWS (where they could use DynamoDB) or GCP (where they could use Firestore) must be willing and capable to operate [etcd](https://etcd.io/), as that's the only HA-supporting, self-runnable backend currently supported by Teleport. With the Postgres backend, users with infrastructure on Azure will be able to use the managed [Azure Database for PostgreSQL](https://azure.microsoft.com/en-us/products/postgresql/) service, with [Managed Identities](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/) allowing easy and secure authentication between Teleport and the database, and users running infrastructure on bare metal will have an option other than etcd for a HA deployment of Teleport.
+
+## Details
+
+WIP: the following is based on the implementation from the tip of `espadolini/pgbk2 `(`392c9ae5d51464bdf571cad249ea262b465217c8`) at the time of writing.
+
+### Non-goals
+
+We don't intend for this backend to be extensible to completely different RDBMS (such as MariaDB or Microsoft SQL Server); support for other Postgres-compatible databases (such as CockroachDB) should be possible in the future, but is not in scope for this RFD.
+
+### General architecture
+
+The Teleport backend abstraction consists of a key-value store (where both key and value are opaque bytestrings), with a per-item optional Time-To-Live (TTL), and a stream of changes to the data. As Postgres provides a way to fetch changes to the data over time through [logical decoding](https://www.postgresql.org/docs/current/logicaldecoding.html) of the Write-Ahead Log, we can tailor the database schema to the exact needs of the backend operations.
+
+As such, the schema consists of a single `kv` table, defined as such:
+```
+CREATE TABLE kv (
+    key bytea PRIMARY KEY,
+    value bytea NOT NULL,
+    expires timestamp,
+    rev uuid NOT NULL
```

Comment:
@xacrimon pointed out that we could use a `bigserial` here, and instead of generating UUIDs on the application side we could just set `rev = DEFAULT` in our writes, and have the new revision be `RETURNED` by the writes.

Reply:
Thinking about this a bit more, we can also make `rev` just a `bigint NOT NULL` and use `txid_current()` either as the default value or in the write queries.

Reply:
I'm not sure how well sequences (as implicitly used by `bigserial`) are preserved by database dumps and restores, but I don't believe that `txid_current()` is going to work correctly across database replicas or backups, so I think that the random UUID approach is still the better one.